[PT-D][Sharded Tensor] new init api for local tensor and sharding spec auto inference

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72733

To improve the perf cost due to communication in the process of init the sharded tensor. There are two changes in this PR/diff:

1. We create a new API named `_init_from_local_tensor` so that if we have only one local tensor, we can initiate a sharded tensor directly from it. (GH issue: https://github.com/pytorch/pytorch/issues/72092)
2. We create a new API to infer the sharding spec from global meta data, so we don't have to manually set the sharding spec when it's not `EnumerableShardingSpec`. (GH issue: https://github.com/pytorch/pytorch/issues/67244)

Differential Revision: [D34132739](https://our.internmc.facebook.com/intern/diff/D34132739/)

Commit:
[PT-D][Sharded Tensor] new init api for local tensor and sharding spec auto inference

To improve the perf cost due to communication in the process of init the sharded tensor. There are two changes in this PR/diff:

1. We create a new API named `_init_from_local_tensor` so that if we have only one local tensor, we can initiate a sharded tensor directly from it. (GH issue: https://github.com/pytorch/pytorch/issues/72092)
2. We create a new API to infer the sharding spec from global meta data, so we don't have to manually set the sharding spec when it's not `EnumerableShardingSpec`. (GH issue: https://github.com/pytorch/pytorch/issues/67244)

Differential Revision: [D34132739](https://our.internmc.facebook.com/intern/diff/D34132739/)

[ghstack-poisoned]

Commit:
Update on "[PT-D][Sharded Tensor] new init api for local tensor and sharding spec auto inference"


To improve the perf cost due to communication in the process of init the sharded tensor. There are two changes in this PR/diff:

1. We create a new API named `_init_from_local_tensor` so that if we have only one local tensor, we can initiate a sharded tensor directly from it. (GH issue: https://github.com/pytorch/pytorch/issues/72092)
2. We create a new API to infer the sharding spec from global meta data, so we don't have to manually set the sharding spec when it's not `EnumerableShardingSpec`. (GH issue: https://github.com/pytorch/pytorch/issues/67244)

Differential Revision: [D34132739](https://our.internmc.facebook.com/intern/diff/D34132739/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -357,15 +359,139 @@ def _init_from_local_shards(
+        init_rrefs=False,
+    ) -> "ShardedTensor":
+        """
+        Initialize a ShardedTensor with only one local tensor with global sharded tensor
```
nit: given global sharded tensor size?

fduwjj replied:
Sure, let me update the comment here.

wanchaol commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -357,15 +359,139 @@ def _init_from_local_shards(
+        size and sharding spec on each rank.
+
+        Args:
+            local_tensor (Tensor): Partial result stored in each rank.
```
nit: local shard tensor stored in each rank? Partial results sounds a bit confusing

fduwjj replied:
Maybe let's call it local tensor from the local shard in each rank?

wanchaol commented on torch/distributed/_shard/sharding_spec/api.py:
```diff
@@ -110,3 +110,62 @@ def __post_init__(self):
+    for shard_metadata in shards_metadata:  # type: ignore[attr-defined]
+        placements.append(shard_metadata.placement)
+        local_offsets = shard_metadata.shard_offsets
+        chunk_offset_list.append(sum(local_offsets))
```
why we need to sum the local_offsets list?

fduwjj replied:
Oh, because for here, we don't know the sharding col, so we just directly sum it. 

The assumption here is that, for chunk sharding case, the offset will always be [0, 0, ..., N, ..., 0]. So if we sum it, we  will get N regardless its sharding col.

wanchaol commented on torch/distributed/_shard/sharding_spec/api.py:
```diff
@@ -110,3 +110,62 @@ def __post_init__(self):
+        if len(shard_dims) != 1:
+            break
+        if not chunk_sharding_dim:
+            chunk_sharding_dim = shard_dims[0]
```
is this saying that if we have `len(shard_dims) == 1` for a single metadata, we can assume this is a chunkshardingspec?

it might not always be the case i guess? i.e. a sharded tensor with global size: (10, 10)
Shard 0: offset: [0, 0], 
Shard 1: offset: [5, 0]
Shard 2: offset: [0, 5]
Shard 3: offset: [5, 5]

This is actually not a chunkShardingSpec but rather a EnumerableShardingSpec?

fduwjj replied:
So, for this case. We will know it's an enum because 
For [5, 5], len(shard_dims) != 1.
For [5, 0] and then [0, 5], chunk_sharding_dim != shard_dims[0]. 

No matter what sequence you have in the sharding spec, we will know it's a enum sharding spec.

wanchaol replied:
I see, yeah make sense.

wanchaol reviewed (commented):
Looks good overall, thanks for doing it! have some questions about the inference of the ChunkShardingSpec

Commit:
Update on "[PT-D][Sharded Tensor] new init api for local tensor and sharding spec auto inference"


To improve the perf cost due to communication in the process of init the sharded tensor. There are two changes in this PR/diff:

1. We create a new API named `_init_from_local_tensor` so that if we have only one local tensor, we can initiate a sharded tensor directly from it. (GH issue: https://github.com/pytorch/pytorch/issues/72092)
2. We create a new API to infer the sharding spec from global meta data, so we don't have to manually set the sharding spec when it's not `EnumerableShardingSpec`. (GH issue: https://github.com/pytorch/pytorch/issues/67244)

Differential Revision: [D34132739](https://our.internmc.facebook.com/intern/diff/D34132739/)

[ghstack-poisoned]

fduwjj commented:
Addressed the feedback from the reviewer.

fduwjj commented:
Looks like the failed test is not relevant to this PR.

wanchaol commented on test/distributed/_shard/sharded_tensor/test_sharded_tensor.py:
```diff
@@ -1781,6 +1785,124 @@ def test_with_rpc_names(self):
+        for idx, placement in enumerate(sharding_spec.placements):
+            shard_size = copy.deepcopy(st_size)
+            offsets = [0] * len(st_size)
+            offsets[chunk_sharding_dim] = min(
```
I guess we will have a easier way to generate the ShardedMetadatas for sharding spec when we have https://github.com/pytorch/pytorch/pull/72130

fduwjj replied:
Yes, that's right. I will send a PR to rebase on your change. Let me add a TODO here.

wanchaol commented on torch/distributed/_shard/sharded_tensor/_ops/linear.py:
```diff
@@ -218,8 +215,17 @@ def _handle_col_wise_sharding(input, world_size, weight, rank, local_shard_t, bi
+    st_size[-1] = weight.size(0)
+    new_sharding_spec = ChunkShardingSpec(
+        dim=-1,
+        placements=weight._sharding_spec.placements
```
nit: lets use the public API: weight.sharding_spec() instead.

fduwjj replied:
sure.

wanchaol commented on torch/distributed/_shard/sharding_spec/api.py:
```diff
@@ -110,3 +110,65 @@ def __post_init__(self):
+def _infer_sharding_spec_from_shards_metadata(shards_metadata):
+    """
+    Infer the sharding spec from the metadata of each shard of a ShardedTensor.
+    If the tensor is sharded only on one dimension, we then assume it's chunk sharded.
```
nit: it's a ChunkShardingSpec?

wanchaol commented on torch/distributed/_shard/sharding_spec/api.py:
```diff
@@ -110,3 +110,65 @@ def __post_init__(self):
+            x[chunk_sharding_dim]
+            for _, x in sorted(zip(chunk_offset_list, shard_size_list))
+        ]
+        if len(shard_sizes) == 1 or (
```
can you also make a comment to explain what this logic used for?

fduwjj replied:
sure, this is for corner cases like when enum sharding will also shard by one dimension but the offset is not the same, aka, the tensor is not evenly distributed.

wanchaol reviewed (approved):
lgtm, thanks!

Commit:
Update on "[PT-D][Sharded Tensor] new init api for local tensor and sharding spec auto inference"


To improve the perf cost due to communication in the process of init the sharded tensor. There are two changes in this PR/diff:

1. We create a new API named `_init_from_local_tensor` so that if we have only one local tensor, we can initiate a sharded tensor directly from it. (GH issue: https://github.com/pytorch/pytorch/issues/72092)
2. We create a new API to infer the sharding spec from global meta data, so we don't have to manually set the sharding spec when it's not `EnumerableShardingSpec`. (GH issue: https://github.com/pytorch/pytorch/issues/67244)

Differential Revision: [D34132739](https://our.internmc.facebook.com/intern/diff/D34132739/)

[ghstack-poisoned]
